### PR TITLE
Use keyspace info for replication factor

### PR DESCRIPTION
**Pull request by conorbros**
This PR gets the keyspace metadata from the `system_schema.keyspaces` table. When a keyspace using `NetworkTopologyStrategy` we only use the replication factor for the data center the Shotover node is in. Processing schema change events occurs the same as topology change events.

### CODE REVIEW

File: shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs [hunk 1]
```diff
@@ -131,19 +145,37 @@ impl NodePool {
                 .clone()
         };
 
+        let keyspace = {
```
Comment (conorbros) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:148]: Need to handle this error better, we should just do normal routing if we don't have the keyspace info.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs [hunk 2]
```diff
@@ -131,19 +145,37 @@ impl NodePool {
...
+                )
+                .ok_or(GetReplicaErr::NoMetadata)?
+                .clone()
+        };
+
         let routing_key = calculate_routing_key(
             &metadata.pk_indexes,
             execute.query_parameters.values.as_ref().ok_or_else(|| {
-                GetReplicaErr::Other(anyhow!("Execute body does not have query paramters"))
+                GetReplicaErr::Other(anyhow!("Execute body does not have query parameters"))
```
Comment (conorbros) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:167]: Same as above.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs [hunk 3]
```diff
@@ -30,6 +30,14 @@ impl TokenMap {
             .take(replica_count)
             .map(|(_, node)| *node)
     }
+
+    // pub fn iter_replica_nodes(&self, token: Murmur3Token) -> impl Iterator<Item = Uuid> + '_ {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs:34]: commented code

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs:34]: Yeah I was just leaving that around in case we needed it but I'll remove it.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs [hunk 4]
```diff
@@ -47,7 +55,7 @@ mod test_token_map {
         vec![
             CassandraNode::new(
                 "127.0.0.1:9042".parse().unwrap(),
-                "rack1".into(),
+                "dc1".into(),
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs:58]: this arg is still for the rack right?

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/token_map.rs:58]: Oh I don't know how I got that mixed up. Thanks

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 5]
```diff
@@ -120,18 +150,31 @@ async fn topology_task_process(
...
                                     }
                                 }
+                                if let Err(watch::error::SendError(_)) =
+                                    nodes_tx.send(nodes.clone())
+                                {
+                                    break 'listen;
+                                }
+                            }
+                            ServerEvent::SchemaChange(change) => {
+                                tracing::warn!("{:?}", change);
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:160]: stray log

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 6]
```diff
@@ -106,9 +124,21 @@ async fn topology_task_process(
                                     }
 
                                     nodes = new_nodes;
+
+                                    if let Err(watch::error::SendError(_)) =
+                                        nodes_tx.send(nodes.clone())
+                                    {
+                                        break 'listen;
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:131]: Can we just immediately `return Ok(())` instead of breaking out and then relying on the return that occurs later on?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 7]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+
+    pub async fn query(
+        connection: &CassandraConnection,
+        data_center: &str,
+    ) -> Result<HashMap<String, KeyspaceMetadata>> {
+        let (tx, rx) = oneshot::channel();
+
+        connection.send(Message::from_frame(
+            Frame::Cassandra(CassandraFrame{
+                version: Version::V4,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:253]: The topology task knows what version the protocol is using via the `let version = connection_info.connection_factory.get_version()?;` line.
We should use that version here.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 8]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+                stream_id: 0,
+                tracing: Tracing::Request(false),
+                warnings: vec![],
+                operation: CassandraOperation::Query{
+                    query: Box::new(
+                        parse_statement_single(
+
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:260]: stray line?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 9]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+    fn into_keyspaces(
+        mut response: Message,
+        data_center: &str,
+    ) -> Result<HashMap<String, KeyspaceMetadata>> {
+        if let Some(Frame::Cassandra(frame)) = response.frame() {
+            match &mut frame.operation {
+                CassandraOperation::Result(CassandraResult::Rows { rows, .. }) => rows
+                    .iter_mut()
+                    .map(|row| build_keyspace(row, data_center))
+                    .try_collect(),
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:284]: both of the `try_collect`s can just be `collect`s.
`try_collect` just helps with type inference sometimes and regular `collect` has its type inferred just fine here.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 10]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+            Frame::Cassandra(CassandraFrame{
+                version: Version::V4,
+                stream_id: 0,
+                tracing: Tracing::Request(false),
+                warnings: vec![],
+                operation: CassandraOperation::Query{
+                    query: Box::new(
+                        parse_statement_single(
+
+                            "SELECT keyspace_name, toJson(replication) AS replication FROM system_schema.keyspaces",
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:261]: Why do we use toJson instead of just letting it send the map over the cassandra protocol?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 11]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+
+        Ok((name, metadata))
+    }
+
+    fn build_replication_strategy(
+        mut properties: Map<String, JsonValue>,
+    ) -> Result<ReplicationStrategy> {
+        match properties.remove("class") {
+            Some(JsonValue::String(class)) => Ok(match class.as_str() {
+                "org.apache.cassandra.locator.SimpleStrategy" | "SimpleStrategy" => {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:346]: Why are there 2 possible values for this?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 12]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+                    operation
+                )),
+            }
+        } else {
+            Err(anyhow!("Failed to parse keyspace query response"))
+        }
+    }
+
+    fn build_keyspace(
+        row: &mut Vec<MessageValue>,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:296]: If we take ownership of row here it alleviates the worry of "oh but what if someone tries to use row after calling this build_keyspace."

image

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 13]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+        data_center: &str,
+    ) -> Result<(String, KeyspaceMetadata)> {
+        let metadata = if let Some(MessageValue::Varchar(string)) = row.pop() {
+            let replication: JsonValue = serde_json::from_str(&string).map_err(|error| {
+                anyhow!(format!(
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:301]: anyhow will do its own formatting (do a quick search for other cases too)

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 14]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+                    )))
+                }
+            };
+
+            let replication_factor = match replication_strategy {
+                ReplicationStrategy::SimpleStrategy { replication_factor } => replication_factor,
+                ReplicationStrategy::NetworkTopologyStrategy {
+                    datacenter_replication_factor,
+                } => *datacenter_replication_factor.get(data_center).unwrap_or(&0),
+                _ => 0,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:322]: Should this be 1 instead of 0?
I imagine 1 means the data is stored in one location?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 15]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+                extract_replication_factor(Some(&replication_factor))
+                    .map(move |replication_factor| (key, replication_factor))
+            })
+            .try_collect()
+    }
+
+    fn extract_replication_factor(value: Option<&JsonValue>) -> Result<usize> {
+        match value {
+            Some(JsonValue::String(replication_factor)) => {
+                let result = if let Some(slash) = replication_factor.find('/') {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:380]: Is there any documentation on this `/` format for the replication_factor?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 16]
```diff
@@ -183,6 +227,174 @@ async fn fetch_current_nodes(
...
+                    usize::from_str(&replication_factor[..slash])
+                } else {
+                    usize::from_str(replication_factor)
+                };
+
+                result.map_err(|error| {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:386]: This should use context 
image

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 17]
```diff
@@ -183,18 +226,187 @@ async fn fetch_current_nodes(
...
+                        )?,
+                    }
+                }
+                "org.apache.cassandra.locator.NetworkTopologyStrategy"
+                | "NetworkTopologyStrategy" => ReplicationStrategy::NetworkTopologyStrategy {
+                    datacenter_replication_factor: extract_datacenter_replication_factor(
+                        properties,
+                    )?,
+                },
+                _ => ReplicationStrategy::Other,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:358]: Is this case just taken when we havent implemented support for a replication strategy yet?
If thats the case we should emit an `error!` and rename `ReplicationStrategy::Other` to `ReplicationStategy::Unknown`

File: shotover-proxy/src/frame/cassandra.rs [hunk 18]
```diff
@@ -205,38 +205,45 @@ impl CassandraFrame {
...
+                                .metadata
+                                .flags
+                                .contains(RowsMetadataFlags::NO_METADATA)
+                            {
+                                rows.rows_content
+                                    .into_iter()
+                                    .map(|row| {
+                                        row.into_iter()
+                                            .map(|row_content| {
+                                                if row_content.is_empty() {
```
Comment (rukai) [inline shotover-proxy/src/frame/cassandra.rs:218]: Can you explain this change?
Shouldnt Null be handled at a lower level?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 14 above)
Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:322]: The answer is that because the replication factor is per dc, we can have a replication factor of 0 in one dc while having a non-zero replication factor in other dc's, so 0 does make semantic sense.

I''m still a little concerned about this line for other reasons but that is addressed in https://github.com/shotover/shotover-proxy/pull/895/files#r1012544156

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 11 above)
Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:346]: I checked driver implementations and noticed they did this, I'd imagine there is some cases where we don't get the fully qualified name.

File: shotover-proxy/src/frame/cassandra.rs (hunk 18 above)
Comment (conorbros) [reply shotover-proxy/src/frame/cassandra.rs:218]: The Datastax driver doesn't seem to let you create unbound prepared statements and this causes that error (tries to unwrap on a null value). I'm seeing if I can fix it by finishing off the scylla test helpers.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 19]
```diff
@@ -183,18 +226,147 @@ async fn fetch_current_nodes(
...
+        connection.send(Message::from_frame(
+            Frame::Cassandra(CassandraFrame{
+                version,
+                stream_id: 0,
+                tracing: Tracing::Request(false),
+                warnings: vec![],
+                operation: CassandraOperation::Query{
+                    query: Box::new(
+                        parse_statement_single(
+                            "SELECT keyspace_name, replication AS replication FROM system_schema.keyspaces",
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:249]: `replication AS replication` -> `replication`

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 17 above)
Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs]: Closing and reopening on a more recent line

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 20]
```diff
@@ -183,18 +226,147 @@ async fn fetch_current_nodes(
...
+                    KeyspaceMetadata {
+                        replication_factor: data_center_rf,
+                    }
+                }
+                "org.apache.cassandra.locator.LocalStrategy" | "LocalStrategy" => {
+                    KeyspaceMetadata {
+                        replication_factor: 0,
+                    }
+                }
+                _ => KeyspaceMetadata {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:339]: If strategies are exhaustive we should return an Err here instead.
If they are not exhaustive because users can make custom ones or something, then we should log an error! here and probably return a replication_factor of 1 as that sounds like a safer default?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 21]
```diff
@@ -183,18 +226,147 @@ async fn fetch_current_nodes(
...
+                        None => 0,
+                    };
+
+                    KeyspaceMetadata {
+                        replication_factor: data_center_rf,
+                    }
+                }
+                "org.apache.cassandra.locator.LocalStrategy" | "LocalStrategy" => {
+                    KeyspaceMetadata {
+                        replication_factor: 0,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:336]: a constant of 0 cant be right?

File: shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs [hunk 22]
```diff
@@ -102,6 +102,12 @@ async fn use_statement(session: &CassandraConnection) {
     .await;
 }
 
+async fn hardcoded_prepared(session: &CassandraConnection) {
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs:105]: This is testing a bug fix that is separate to the main goal of using replication_factor in our routing, can this test and the bug fix be moved into a separate PR?

Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs]: As part of implementing the keyspace info I added error handling which fixes that bug so I included that in this PR.

**Review by benbromhead — APPROVED** [review]

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 11 above)
Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs]: I tracked the source down to this https://github.com/scylladb/scylla-rust-driver/pull/327 but I cant see an appropriate justification for the change.
I also experimented locally with both cassandra and scylla, and I cant find a way to make `system_schema.keyspaces` output in that shortened form.

Unless you have any ideas as to why its there, shall we remove it?

Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs]: Actually, it seems like it gets output in that format when `DESCRIBE KEYSPACE` is used.
Not sure if thats actually possible anymore, it seems to be deprecated and/or DSE specific but its probably worth keeping the shorter form around just in case the code gets repurposed for that use in the future.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs [hunk 23]
```diff
@@ -15,13 +16,20 @@ use tokio::sync::{watch, RwLock};
...
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct KeyspaceMetadata {
+    pub replication_factor: usize,
+}
+
 #[derive(Debug)]
 pub struct NodePool {
     prepared_metadata: Arc<RwLock<HashMap<CBytesShort, PreparedMetadata>>>,
+    keyspace_metadata: Arc<RwLock<HashMap<String, KeyspaceMetadata>>>,
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/node_pool.rs:32]: prepared_metadata uses an `Arc<RwLock<T>>` because each connection needs to be able to mutate a global state because preparation can occur on any connection.
However that same reasoning does not apply to `keyspace_metadata`.
Each connection receives the keyspace_metadata from the topology task and has no need to share this state with other connections because they all received the same state.
Changing this type to a `HashMap<String, KeyspaceMetadata>` should be sufficient to resolve this issue.

File: shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs [hunk 24]
```diff
@@ -119,4 +119,10 @@ where
     select_cross_connection(session, connection_creator).await;
     delete(session).await;
     use_statement(session).await;
+
+    if session.is(&[CassandraDriver::Scylla, CassandraDriver::CdrsTokio]) {
```
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/prepared_statements.rs:123]: The tests pass just fine on my machine when I remove this check.
Did something else incidentally fix the issue we were having with the datastax driver?

**Review by rukai — APPROVED** [review]
LGTM
